### PR TITLE
Added categories

### PR DESCRIPTION
**Pull request by suprun-anna**

### CODE REVIEW

File: src/main/java/mate/academy/bookstore/controller/CategoryController.java [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+    @PreAuthorize("hasRole('ROLE_USER')")
+    @GetMapping
+    public List<CategoryDto> getAll(Pageable pageable) {
+        return categoryService.findAll(pageable);
+    }
+
+    @Operation(summary = "Get category by id",
+            description = "Get category by id")
+    @PreAuthorize("hasRole('ROLE_USER')")
+    @GetMapping("{id}")
```
Comment (Rommelua) [inline src/main/java/mate/academy/bookstore/controller/CategoryController.java]:
```suggestion
    @GetMapping("/{id}")
```

File: src/main/java/mate/academy/bookstore/controller/CategoryController.java [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+    public CategoryDto getCategoryById(@PathVariable Long id) {
+        return categoryService.getById(id);
+    }
+
+    @Operation(summary = "Get books by category", description = "Get books by category")
+    @PreAuthorize("hasRole('ROLE_USER')")
+    @GetMapping("{id}/books")
```
Comment (Rommelua) [inline src/main/java/mate/academy/bookstore/controller/CategoryController.java]:
```suggestion
    @GetMapping("/{id}/books")
```

File: src/main/java/mate/academy/bookstore/model/Book.java [hunk 3]
```diff
@@ -29,6 +35,11 @@ public class Book {
     private BigDecimal price;
     private String description;
     private String coverImage;
+    @ManyToMany(fetch = FetchType.EAGER)
```
Comment (Rommelua) [inline src/main/java/mate/academy/bookstore/model/Book.java]: use Lasy

File: src/main/java/mate/academy/bookstore/service/impl/BookServiceImpl.java [hunk 4]
```diff
@@ -52,17 +60,26 @@ public BookDto getById(Long id) {
...
+    @Override
+    public BookDtoWithoutCategoryIds getBookWithoutCategoryById(Long id) {
+        Book book = bookRepository.findById(id).orElseThrow(
+                () -> new EntityNotFoundException("Can't find book by id=" + id));
+        return bookMapper.toDtoWithoutCategories(book);
+    }
+
     @Override
     public BookDto updateBookById(Long id, CreateBookRequestDto bookDto) {
         Optional<Book> optionalBook = bookRepository.findById(id);
```
Comment (Rommelua) [inline src/main/java/mate/academy/bookstore/service/impl/BookServiceImpl.java]:
```suggestion
        Book book = bookRepository.findById(id).orElseThrow(...);
```

File: src/main/java/mate/academy/bookstore/service/impl/BookServiceImpl.java [hunk 5]
```diff
@@ -52,17 +60,26 @@ public BookDto getById(Long id) {
...
-            book.setPrice(bookDto.getPrice());
-            book.setDescription(bookDto.getDescription());
-            book.setCoverImage(bookDto.getCoverImage());
+            book.setTitle(bookDto.title());
+            book.setAuthor(bookDto.author());
+            book.setIsbn(bookDto.isbn());
+            book.setPrice(bookDto.price());
+            book.setDescription(bookDto.description());
+            book.setCoverImage(bookDto.coverImage());
+            Set<Category> categories = getCategoriesByIds(bookDto.categoryIds());
```
Comment (Rommelua) [inline src/main/java/mate/academy/bookstore/service/impl/BookServiceImpl.java]: use mapper for this work

File: src/main/java/mate/academy/bookstore/service/impl/CategoryServiceImpl.java [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+
+    @Override
+    public CategoryDto save(CategoryDto categoryDto) {
+        Category category = categoryMapper.toEntity(categoryDto);
+        return categoryMapper.toDto(categoryRepository.save(category));
+    }
+
+    @Override
+    public CategoryDto update(Long id, CategoryDto categoryDto) {
+        Optional<Category> optionalCategory = categoryRepository.findById(id);
```
Comment (Rommelua) [inline src/main/java/mate/academy/bookstore/service/impl/CategoryServiceImpl.java]: orElseThrow

File: src/main/java/mate/academy/bookstore/service/impl/CategoryServiceImpl.java [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+        if (optionalCategory.isPresent()) {
+            Category category = optionalCategory.get();
+            category.setName(categoryDto.getName());
+            category.setDescription(categoryDto.getDescription());
```
Comment (Rommelua) [inline src/main/java/mate/academy/bookstore/service/impl/CategoryServiceImpl.java]: use mapper to update

File: src/main/resources/db/changelog/changes/08-create-categories-table.yaml [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+                  type: VARCHAR(255)
+              - column:
+                  name: is_deleted
+                  type: BOOLEAN
+                  defaultValue: false
+                  constraints:
+                    nullable: false
+      rollback:
+        - dropTable:
+            tableName: categories
```
Comment (Vasul-Pankiv) [inline src/main/resources/db/changelog/changes/08-create-categories-table.yaml:32]: File has to end with empty line

File: src/main/resources/db/changelog/changes/09-create-books-categories-table.yaml [hunk 9]
```diff
@@ -0,0 +1,21 @@
...
+                  constraints:
+                    primaryKey: true
+              - column:
+                  name: category_id
+                  type: BIGINT
+                  constraints:
+                    primaryKey: true
+      rollback:
+        - dropTable:
+            tableName: user_roles
```
Comment (Vasul-Pankiv) [inline src/main/resources/db/changelog/changes/09-create-books-categories-table.yaml:21]: same

**Review by Vasul-Pankiv — APPROVED** [review]
Good job

**Pull request merged**